item/historical/scripts/T002.py

```diff
@@ -0,0 +1,112 @@
+"""Data cleaning code and configuration for T002."""
+from .util.managers.dataframe import ColumnName
+from functools import lru_cache
+import pandas as pd
+
+#: Dimensions and attributes which do not vary across this data set.
+COMMON_DIMS = dict(
+    # Add the same source to all rows since all data comes from the same source
+    source="International Transport Forum",
+    # Since all the data is associated to "Freight," the Service is "Freight"
+    service="Freight",
+    # The dataset does not provide any data on the following columns, so we
+    # add the default value of "All" in both cases
+    technology="All",
+    fuel="All",
+    vehicle_type="Container",
+)
+
+#: Columns to drop from the raw data.
+COLUMNS = dict(
+    drop=[
+        "COUNTRY",
+        "VARIABLE",
+        "YEAR",
+        "Unit Code",
+        "PowerCode Code",
+        "PowerCode",
+        "Reference Period Code",
+        "Reference Period",
+        "Flag Codes",
+        "Flags",
+    ],
+    # Column containing country name for determining ISO 3166 alpha-3 codes and
+    # iTEM regions. Commented, because this is the default value.
+    # country_name='Country',
+)
+
+
+def process(df):
+    """Process data set T002."""
+    # Getting a generic idea of what countries are missing values and dropping
+    # NaN values
+    #
+    # Rule: Erase all value with NaN
+
+    list_of_countries_with_missing_values = list(
+        set(df[df["Value"].isnull()]["Country"])
+    )
+    print(
+        ">> Number of countries missing values: {}".format(
+            len(list_of_countries_with_missing_values)
+        )
+    )
+    print(">> Countries missing values:")
+    print(list_of_countries_with_missing_values)
+    print(">> Number of rows to erase: {}".format(len(df[df["Value"].isnull()])))
+
+    # Assign the 'Mode' and 'Variable' columns
+    df = pd.concat([df, df["Variable"].apply(mode_and_variable)], axis=1)
+
+    # TODO Find a way on dealing with the "Variable" columns case in a better way
+    # When the value for the "Variable" column is not unique, the code breaks.
+    del df["Variable"]
+    df.rename({"Var 2": ColumnName.VARIABLE.value}, axis=1, inplace=True)
+
+    # Assign the 'Unit' column
+    df = pd.concat([df, df["Unit"].apply(unit)], axis=1)
+
+    # TODO same as in 'Variable' column
+    del df["Unit"]
+    df.rename({"Unit 2": ColumnName.UNIT.value}, axis=1, inplace=True)
+
+    # 1. Drop null values.
+    # 2. Convert to the preferred iTEM units.
+    df = df.dropna()
+
+    return df
+
+
+@lru_cache()
+def mode_and_variable(variable_name):
+
+    # Mode
+    if "Rail" in variable_name:
+        mode = "Rail"
+    else:
+        mode = "Shipping"
+
+    # Variable
+    if "TEU" in variable_name:
+        variable = "Freight (TEU)"
+    else:
+        variable = "Freight (Weight)"
+
+    return pd.Series(
+        [mode, variable],
+        index=[ColumnName.MODE.value, "Var 2"],
```

Comment:
@khaeru 
Problem: The raw dataset has a column named "Variable". I cannot put such column in the list `drop` because I need it to determine "Mode" and "Variable" columns. However, when I do want to add the "real" column named "Variable," I end up having two columns with the same name and then unable to determine which one to erase.

My solution so far: As seen in this line, I set a temporary name for the "real" column named "Variable" and then in lines 63 and 64, I delete the unwanted "Variable" column and then renamed "Var 2" to "Variable", respectively

Question: I am not sure if this is the most appropriate manner of handling this scenario. What would you suggest to implement?

Note: This same name collision problem happens with the "Units" column.

Reply:
This seems like it would work. A general recommendation: use pandas method chaining as much as possible.

An equivalent to:
```python
del df["Unit"]
```
is 
```python
df.drop(columns="Unit")
```

So this means that instead of:
```python
df = pd.concat([df, df["Variable"].apply(mode_and_variable)], axis=1)
del df["Variable"]
df.rename({"Var 2": ColumnName.VARIABLE.value}, axis=1, inplace=True)
```
…you can do:
```python
df = (
    pd.concat([df, df["Variable"].apply(mode_and_variable)], axis=1)
    .drop(columns="Variable")
    .rename({"Var 2": ColumnName.VARIABLE.value}, axis=1, inplace=True)
)	
```
This is better idiom because pandas can streamline these chained calls and, often, execute them more efficiently.

However, it seems like you could also do this:
```python
df = pd.concat(
  [df.drop(columns="Variable"), df["Variable"].apply(mode_and_variable)],
  axis=1,
)
```
i.e. you use the "Variable" column to `apply()` the helper function, and concatenate the results to a data frame from which "Variable" is *already dropped*. This approach has the benefit of not introducing new column names like "Var 2" that need to be explained.